executor.go

```diff
@@ -48,9 +51,160 @@ type ValidationResult struct {
+	if err != nil {
+		return nil, fmt.Errorf("failed to generate and aggregate verifier reports: %w", err)
+	}
+
+	if e.PolicyEnforcer == nil {
+		return &ValidationResult{
+			Succeeded:       false,
+			ArtifactReports: aggregatedVerifierReports,
+		}, nil
+	}
+
+	decision, err := e.PolicyEnforcer.Evaluate(ctx, aggregatedVerifierReports)
+	if err != nil {
+		return nil, fmt.Errorf("failed to evaluate verifier reports: %w", err)
+	}
+
+	return &ValidationResult{
+		Succeeded:       decision,
+		ArtifactReports: aggregatedVerifierReports,
+	}, nil
+}
+
+// aggregateVerifierReports generates and aggregates all verifier reports.
+func (e *Executor) aggregateVerifierReports(ctx context.Context, opts ValidateArtifactOptions) ([]*ValidationReport, error) {
+	var aggregatedVerifierReports []*ValidationReport
+	var err error
+
+	// TODO: Implement a worker pool to validate artifacts concurrently.
+	// TODO: Enforce check on the stack size.
+	taskStack := make([]*task, 0)
+
+	// Iterate over all stores to create tasks for each store so that further
+	// referrer artifacts will be fetched from the same store.
+	for _, store := range e.Stores {
+		taskStack = append(taskStack, &task{
+			store:    store,
+			artifact: opts.Subject,
+		})
+	}
+
+	for len(taskStack) > 0 {
+		task := taskStack[len(taskStack)-1]
+		taskStack = taskStack[:len(taskStack)-1]
+
+		taskStack, aggregatedVerifierReports, err = e.processTask(ctx, task, taskStack, aggregatedVerifierReports, opts.ReferenceTypes)
+		if err != nil {
+			return nil, fmt.Errorf("failed to validate artifact %s: %w", task.artifact, err)
+		}
+	}
+
+	return aggregatedVerifierReports, nil
+}
+
+// processTask processes the task and returns the updated stack and aggregated
+// verifier reports.
+func (e *Executor) processTask(ctx context.Context, task *task, stack []*task, aggregatedVerifierReports []*ValidationReport, referenceTypes []string) ([]*task, []*ValidationReport, error) {
+	artifactRef := task.artifact
+	ref, err := registry.ParseReference(artifactRef)
+	if err != nil {
+		return stack, nil, fmt.Errorf("failed to parse artifact reference %s: %w", artifactRef, err)
+	}
+
+	artifactDesc, err := task.store.Resolve(ctx, ref.String())
+	if err != nil {
+		return stack, nil, fmt.Errorf("failed to resolve artifact reference %s: %w", ref.Reference, err)
+	}
+	ref.Reference = artifactDesc.Digest.String()
+
+	stack, artifactReports, err := e.verifyAgainstReferrers(ctx, task.store, ref, stack, referenceTypes)
+	if err != nil {
+		return stack, nil, err
+	}
+
+	// If the current task is the root task, add the artifactReports to the
+	// aggregatedVerifierReports. Otherwise, they are just artifactReports of
+	// the subject artifact.
+	if task.subjectReport != nil {
+		task.subjectReport.ArtifactReports = append(task.subjectReport.ArtifactReports, artifactReports...)
+	} else {
+		aggregatedVerifierReports = append(aggregatedVerifierReports, artifactReports...)
+	}
```

Comment:
Here we have some edge cases. What if the same artifact is there in different stores? Will they be aggregated?